### PR TITLE
feat: Add configurable batch_size and max_workers to embed method

### LINKED ISSUE #534: Allow users to configure `embed_batch_size` or `ThreadPoolExecutor` size when calling `Client.embed`
It looks like batching was added in #437 - thank you for implementing this, it's very helpful.

I notice that batching, as defined [here](https://github.com/cohere-ai/cohere-python/blob/main/src/cohere/client.py#L135), depends on a [fixed batch size](https://github.com/cohere-ai/cohere-python/blob/main/src/cohere/config.py#L1). This can be suboptimal for clients submitting a large number of smaller documents, as we cannot [configure the ThreadPoolExecutor size](https://github.com/cohere-ai/cohere-python/blob/main/src/cohere/client.py#L111) to parallelize a large number of small data payloads. As a result a client might end up blocking while waiting for small network responses.

Would it be possible to allow clients to configure either the `ThreadPoolExecutor` size or the `embed_batch_size` setting when calling `embed`?

### PR DESCRIPTION
**Pull request by fede-kamel**
# Add configurable batch_size and max_workers to embed method

## Summary
This PR fixes #534 by making the embed batch size configurable through optional parameters, giving users control over batching behavior based on their specific needs.

## Problem
Previously, the `embed()` method used a fixed batch size of 96 (from `config.embed_batch_size`), which could be suboptimal for various use cases:
- Users with memory constraints needed smaller batches
- Users with high-throughput needs wanted larger batches
- Rate-limited applications needed to control concurrency

## Solution
Added two optional parameters to the `embed()` method:
- `batch_size: Optional[int] = None` - Controls the number of texts per batch
- `max_workers: Optional[int] = None` - Controls ThreadPoolExecutor concurrency (sync client only)

### Changes to `src/cohere/client.py`:
```python
def embed(
    self,
    *,
    texts: Optional[Sequence[str]] = OMIT,
    # ... existing parameters ...
    batch_size: Optional[int] = None,  # NEW
    max_workers: Optional[int] = None,  # NEW
) -> EmbedResponse:
```

The implementation:
1. Uses provided `batch_size` or falls back to the default `embed_batch_size` (96)
2. Creates a temporary ThreadPoolExecutor if `max_workers` is specified
3. Maintains full backward compatibility - existing code continues to work unchanged

### All tests pass:
```bash
$ python -m pytest tests/test_configurable_batch_size.py -v
============================= test session starts ==============================
collected 6 items

tests/test_configurable_batch_size.py::TestConfigurableBatchSize::test_batch_size_edge_cases PASSED [ 16%]
tests/test_configurable_batch_size.py::TestConfigurableBatchSize::test_custom_batch_size PASSED [ 33%]
tests/test_configurable_batch_size.py::TestConfigurableBatchSize::test_custom_max_workers PASSED [ 50%]
tests/test_configurable_batch_size.py::TestConfigurableBatchSize::test_default_batch_size PASSED [ 66%]
tests/test_configurable_batch_size.py::TestConfigurableBatchSize::test_no_batching_ignores_parameters PASSED [ 83%]
tests/test_configurable_batch_size.py::TestAsyncConfigurableBatchSize::test_async_custom_batch_size PASSED [100%]

============================== 6 passed in 0.40s ===============================
```

### Test coverage includes:
- ✅ Custom batch sizes work correctly
- ✅ Default batch size (96) is used when parameter not specified
- ✅ Edge cases: batch_size=1, batch_size > total texts
- ✅ Custom max_workers creates new ThreadPoolExecutor
- ✅ Parameters are properly ignored when batching=False
- ✅ Async client batch_size support

## Code Quality
- ✅ Ruff linting passes
- ✅ Mypy type checking passes
- ✅ Import ordering fixed automatically by ruff

### Default behavior (unchanged):
```python
response = client.embed(texts=texts, model="embed-english-v3.0")
# Uses default batch_size=96
```

### Custom batch size for memory optimization:
```python
response = client.embed(
    texts=texts,
    model="embed-english-v3.0", 
    batch_size=10  # Smaller batches for memory-constrained environments
)
```

### Rate limiting with reduced concurrency:
```python
response = client.embed(
    texts=texts,
    model="embed-english-v3.0",
    batch_size=20,
    max_workers=2  # Only 2 concurrent API calls
)
```

## Benefits
1. **Memory optimization**: Users can reduce batch size to limit memory usage
2. **Performance tuning**: Users can increase batch size for fewer API calls
3. **Rate limit handling**: Control concurrency with max_workers
4. **Backward compatible**: No changes required to existing code
5. **Complements PR #698**: Works well with the memory-efficient `embed_stream()` method

This implementation provides the flexibility requested in issue #534 while maintaining the SDK's ease of use and backward compatibility.

---

> [!NOTE]
> **Enable memory‑efficient embedding and configurable batching**
> 
> - Adds `embed_stream` to `BaseCohere` and `v2` clients to stream embeddings per-item, batching inputs and parsing incrementally via `StreamingEmbedParser` in `streaming_utils.py` (uses `ijson` when available, falls back to JSON)
> - Introduces `StreamedEmbedding` data class and helpers for incremental parsing of `embeddings_floats` and `embeddings_by_type`
> - Extends `Client.embed`/`AsyncClient.embed` with optional `batch_size`; sync version also supports `max_workers` (temporary `ThreadPoolExecutor` with cleanup). Validates `batch_size`; async warns and ignores `max_workers`
> - Updates tests: configurable batch size (sync/async), streaming behavior and memory efficiency; adds integration-style examples. Minor housekeeping: add `.venv/` to `.gitignore`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 792b57e23c2f40f3d452bfcca1c06a597849c229. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by fede-dash** [comment]
## Context: How this PR relates to #536 and issue #534

I noticed that PR #536 was already merged, which partially addressed issue #534 by adding  configuration to the Client constructor. After analyzing both implementations, I believe this PR (#699) is still valuable as it complements #536 by addressing the remaining requirements from issue #534.

### What PR #536 provided:
- Client-level ThreadPoolExecutor configuration via constructor
- Example: `client = cohere.Client(thread_pool_executor=ThreadPoolExecutor(32))`

### What this PR adds:
1. **Configurable batch_size** - The other key request from issue #534 that wasn't addressed
2. **Per-call flexibility** - Configure batch_size and max_workers for individual embed() calls
3. **Dynamic optimization** - Adjust parameters based on document characteristics without recreating the client

### Key differences:
| Aspect | PR #536 | This PR (#699) |
|--------|---------|----------------|
| Configuration level | Client-wide | Per-method call |
| Parameters | thread_pool_executor (constructor) | batch_size, max_workers (embed method) |
| Use case | Set once for all operations | Dynamic per-operation tuning |
| Batch size control | ❌ | ✅ |

### Example usage showing both PRs working together:
```python
# PR #536 - Set default thread pool for client
client = cohere.Client(thread_pool_executor=ThreadPoolExecutor(32))

# PR #699 - Override for specific operations
# Small documents: smaller batches, more workers
response = client.embed(texts=small_docs, batch_size=10, max_workers=64)

# Large documents: larger batches, fewer workers  
response = client.embed(texts=large_docs, batch_size=50, max_workers=8)

# Memory constrained: very small batches
response = client.embed(texts=texts, batch_size=5)
```

This implementation completes the solution for issue #534 by providing both the batch size configuration and per-call flexibility that users requested for optimizing their embedding workflows.

**Comment by fede-kamel** [comment]
Hi @mkozakov, @billytrend-cohere, @daniel-cohere! 👋 

Dudes come on!

**Comment by fede-kamel** [comment]
Hey @andrewbcohere, no worries at all - totally understand! Just rebased onto the latest main (now includes SDK regeneration through Nov 10th). All unit tests pass. The PR is ready for review whenever you get a chance. Really appreciate you taking the time to look at this!

File: src/cohere/client.py [hunk 1]
```diff
@@ -203,23 +204,34 @@ def embed(
...
-                    input_type=input_type,
-                    embedding_types=embedding_types,
-                    truncate=truncate,
-                    request_options=request_options,
-                ),
-                texts_batches,
-            )
-        ]
+        effective_batch_size = batch_size if batch_size is not None else embed_batch_size
+        texts_batches = [textsarr[i : i + effective_batch_size] for i in range(0, len(textsarr), effective_batch_size)]
```
[the remark this answers is not in the document]
Comment (fede-kamel) [reply src/cohere/client.py:212]: Fixed - added validation at the start of both sync and async `embed()` methods:

```python
if batch_size is not None and batch_size < 1:
    raise ValueError("batch_size must be at least 1")
```

File: src/cohere/base_client.py [hunk 2]
```diff
@@ -1125,6 +1125,110 @@ def embed(
...
+                request_options=request_options,
+            )
+            
+            # Parse embeddings from response incrementally
+            parser = StreamingEmbedParser(response._response, batch_texts)
+            for i, embedding in enumerate(parser.iter_embeddings()):
+                # Adjust index for global position
+                embedding.index = batch_start + i
+                embedding.text = texts_list[embedding.index]
+                yield embedding
```
[the remark this answers is not in the document]
Comment (fede-kamel) [reply src/cohere/base_client.py:1243]: Fixed - the parser now correctly tracks text index per embedding type. The `embed_stream` methods use the text from the parser to determine the correct global index:

```python
for embedding in parser.iter_embeddings():
    if embedding.text and embedding.text in batch_texts:
        text_idx_in_batch = batch_texts.index(embedding.text)
        embedding.index = batch_start + text_idx_in_batch
    yield embedding
```

File: src/cohere/streaming_utils.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        response_content = self.response.content
+
+        try:
+            # Use ijson for memory-efficient parsing
+            parser = ijson.parse(io.BytesIO(response_content))
+            yield from self._parse_with_ijson(parser)
+        except Exception:
+            # If ijson parsing fails, fallback to regular parsing using buffered content
+            data = json.loads(response_content)
+            yield from self._iter_embeddings_fallback_from_dict(data)
```
[the remark this answers is not in the document]
Comment (fede-kamel) [reply src/cohere/streaming_utils.py:79]: Fixed - now collect all ijson embeddings into a list first, and only yield after successful complete parsing. If ijson fails mid-parse, we reset `embeddings_yielded` and use the fallback on the buffered content:

```python
try:
    embeddings = list(self._parse_with_ijson(parser))
    yield from embeddings  # Only yield after success
except Exception:
    self.embeddings_yielded = 0  # Reset counter
    data = json.loads(response_content)
    yield from self._iter_embeddings_fallback_from_dict(data)
```

File: src/cohere/client.py [hunk 4]
```diff
@@ -394,9 +412,20 @@ async def embed(
...
 
+        textsarr: typing.Sequence[str]  = texts if texts is not OMIT and texts is not None else []
+        effective_batch_size = batch_size if batch_size is not None else embed_batch_size
+        texts_batches = [textsarr[i : i + effective_batch_size] for i in range(0, len(textsarr), effective_batch_size)]
+
+        # Note: max_workers parameter is not used in async version since asyncio.gather
+        # handles concurrency differently than ThreadPoolExecutor
+        if max_workers is not None:
+            # Log a warning or silently ignore - asyncio manages its own concurrency
+            pass
```
[the remark this answers is not in the document]
Comment (fede-kamel) [reply src/cohere/client.py]: Fixed in commit 7c198eaa - Now raises explicit UserWarning when max_workers is used with AsyncClient, explaining that the parameter is not applicable since asyncio.gather() manages concurrency automatically.

File: src/cohere/base_client.py [hunk 5]
```diff
@@ -1125,6 +1125,111 @@ def embed(
...
+            )
+            
+            # Parse embeddings from response incrementally
+            parser = StreamingEmbedParser(response._response, batch_texts)
+            for embedding in parser.iter_embeddings():
+                # The parser sets embedding.text correctly for multiple embedding types
+                # Adjust the global index based on text position in batch
+                if embedding.text and embedding.text in batch_texts:
+                    text_idx_in_batch = batch_texts.index(embedding.text)
+                    embedding.index = batch_start + text_idx_in_batch
```
[the remark this answers is not in the document]
Comment (fede-kamel) [reply src/cohere/base_client.py]: Fixed in commit 73545e56 - Now tracks used indices to handle duplicate texts correctly. Each duplicate text receives its correct sequential index instead of all getting the first occurrence's index.

Example with duplicates:
- Input: [\hello\, \world\, \hello\]  
- Before: indices [0, 1, 0] - WRONG
- After: indices [0, 1, 2] - CORRECT
